### ISSUE TITLE
MEP_oM: Refactor MEP Elements

### ISSUE THREAD

**Issue by michaelhoehn** [BHoM/BHoM#1139]
#### Description:

It has been suggested that the current MEP linear objects would benefit from a bit of refactoring to align more with other more generic Physical objects. Currently, MEP objects have a sectionProperty that contains loads of things pertaining to both geometric and physical, material properties. 

Then each object has it's own sectionProperties (i.e. `DuctSectionProperty`, `PipeSectionProperty`, `CableTraySectionProperty` and so on). This could be much simpler by allowing the objects to simply host `Construction` and `ShapeProfile` properties as demonstrated below. 

![image](https://user-images.githubusercontent.com/54593392/103576791-adf2be00-4ea1-11eb-8ee8-2caabfaddbf0.png)

Although this organisation would help in any physical-based evaluation methods, I still struggle with the potential for conflicting data. For example, a ShapeProfile's thickness could differ from the prescribed Construction layer thickness. So not sure how to best resolve that moving forward. 

Other thoughts are welcome! 

FYI: @FraserGreenroyd @kayleighhoude @IsakNaslundBh

**Comment by rolyhudson** [comment]
Just noting here that if there is a goal of doing topological queries (both spatial and non-spatial) on MEP_oM using new graph tools, life could be made easier if the StartPoint and EndPoint were not `Point` objects but some kind of `IBHoMObject`. 

Why is that? `IBHoMObject` gives us a `BHoM_Guid` which is used to define really simple relationships between objects and `IBHoMObject` allows attaching topological fragments that can be used to specify these relationships.

I know this would allow the existing shortest path methods to work across any MEP system network which has been a use case of interest for some time. 

Also I believe this could also make life easier for pulling networks of these systems in from external sources as the graph create methods have been set up to handle defining the topology. This has been discussed with @pawelbaran @tiagogrossi.

If this is out of scope for this issue please ignore and accept an apology for reading this far. Given the intent of the title I'm mentioning this here.

**Comment by pawelbaran** [comment]
I strongly agree with @rolyhudson - I would mention one more reason for implementing discipline-specific MEP nodes instead of geometrical primitives: besides handling of topological relationships mentioned above, there is also an engineering layer that may require storing more information about MEP element's end point than simply its location. Please correct me if I am truly wrong - otherwise I would consider implementing a discipline-specific node _just in case_ such feature is needed at some point.

**Comment by tiagogrossi** [comment]
Not much to add that hasn't been said, so +1 from me :)

**Comment by IsakNaslundBh** [comment]
Would also agree on the addition of an MEP node. From a usability standpoint, can always add in explicit casting from point, as has been done for structures:

https://github.com/BHoM/BHoM/blob/ba4e0bddcf8bb86497b2e10f1663ab2f31319217/Structure_oM/Elements/Node.cs#L49-L67

With that, it is possible to provide a point (or cartesian coordinate system as it stands for structures) to any component in the UI that expects a Node.

As for the profile refinement/simplification, I do agree that it could be made more simple, but would personally not use the Construction class made for the physical 2D elements.

One thing you could explore is to only store the base profile and the lining and insulation thicknesses, and have the profiles for the latter two be query-able instead. That way you end up with one less layer of objects:

![image](https://user-images.githubusercontent.com/22005920/104179340-f8c58600-540b-11eb-9c57-c6c580a529c2.png)

(not sure the image above is that clear, but happy to have a chat if you like)

**Comment by tiagogrossi** [comment]
Chaps, would it be worthwhile to put together a task force to action this forward? @michaelhoehn ?

**Comment by FraserGreenroyd** [comment]
@tiagogrossi discrete actions will be discussed and communicated during sprint coordination later today if that's what you're asking?

**Comment by tiagogrossi** [comment]
Ah exactly @FraserGreenroyd ! Would be so kind to send the invite to me again please?

**Comment by michaelhoehn** [comment]
Hey @tiagogrossi and others, as mentioned in sprint planning, this refactor will not occur until this PR is closed out https://github.com/BHoM/BHoM/pull/1113.

**Comment by pawelbaran** [comment]
One more item to be considered while resolving this: as pointed by @AndreBalasian, the dimension used in MEP calcs is the nominal dimension, which is usually different from the physical ones, i.e. inner/outer. This means that besides the `SectionProfile` defining the physical dimensions, we also need another property describing the nominal values, i.e. for Pipe I would consider a `NominalDiameter` property.

I hope the above makes sense (particularly to MEP engineers 🙈), please correct me if I am confusing things.

**Comment by travispotterBH** [comment]
To be worked on with @FraserGreenroyd  on secondment potentially  Milestone 6.1 sprint 4

**Comment by travispotterBH** [comment]
Related: https://github.com/BHoM/BHoM/issues/617

**Comment by travispotterBH** [comment]
Related: https://github.com/BHoM/BHoM_Engine/tree/MEP_Engine-%232869-GraphFromMEPObjects
